src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -219,12 +219,139 @@ ss::future<try_abort_reply> tx_gateway_frontend::dispatch_try_abort(
+      shard, _ssg, [pid, tx_seq, timeout](tx_gateway_frontend& self) {
+          auto partition = self._partition_manager.local().get(
+            model::tx_manager_ntp);
+          if (!partition) {
+              vlog(
+                clusterlog.warn,
+                "can't get partition by {} ntp",
+                model::tx_manager_ntp);
+              return ss::make_ready_future<try_abort_reply>(
+                try_abort_reply{.ec = tx_errc::partition_not_found});
+          }
+
+          auto stm = partition->tm_stm();
+
+          if (!stm) {
+              vlog(
+                clusterlog.warn,
+                "can't get tm stm of the {}' partition",
+                model::tx_manager_ntp);
+              return ss::make_ready_future<try_abort_reply>(
+                try_abort_reply{.ec = tx_errc::stm_not_found});
+          }
+
+          return stm->barrier().then([&self, stm, pid, tx_seq, timeout](
+                                       bool ready) {
+              if (!ready) {
+                  return ss::make_ready_future<try_abort_reply>(
+                    try_abort_reply{.ec = tx_errc::unknown_server_error});
+              }
+              auto tx_id_opt = stm->get_id_by_pid(pid);
+              if (!tx_id_opt) {
+                  return ss::make_ready_future<try_abort_reply>(
+                    try_abort_reply{.aborted = true, .ec = tx_errc::none});
+              }
+              auto tx_id = tx_id_opt.value();
+              return stm->get_tx_lock(tx_id)->with(
+                [&self, stm, tx_id, pid, tx_seq, timeout]() {
+                    return self.do_try_abort(stm, tx_id, pid, tx_seq, timeout);
+                });
+          });
+      });
+}
+
+ss::future<try_abort_reply> tx_gateway_frontend::do_try_abort(
+  ss::shared_ptr<tm_stm> stm,
+  kafka::transactional_id transactional_id,
+  model::producer_identity pid,
+  model::tx_seq tx_seq,
+  model::timeout_clock::duration timeout) {
+    auto maybe_tx = co_await stm->get_actual_tx(transactional_id);
+    if (!maybe_tx) {
+        // unknown tx => state was lost => can't be comitted => aborted
+        co_return try_abort_reply{.aborted = true, .ec = tx_errc::none};
+    }
+
+    auto tx = maybe_tx.value();
+
+    if (tx.pid != pid) {
+        // well, that's wierd, it may happen when the coordinator ended
+        // a transaction and initiated a new session for the same tx.id
+        // while the `try_abort` request was inflight
+        co_return try_abort_reply{.ec = tx_errc::request_rejected};
+    }
+
+    if (tx.tx_seq != tx_seq) {
+        // well, that's wierd, it may happen when the coordinator ended
+        // a transaction and started a new one when the `try_abort`
+        // request was inflight
+        co_return try_abort_reply{.ec = tx_errc::request_rejected};
+    }
+
+    if (tx.status == tm_transaction::tx_status::prepared) {
+        co_return try_abort_reply{.commited = true, .ec = tx_errc::none};
```

Comment:
does prepared imply committed (just making sure its not a typo since the other three cases below are all identical)

Reply:
that's true, when we commit a transaction we ask all participating data partitions to prepare and in parallel write preparing status on the transaction coordinator. When all of this succeeds we acknowledge the transaction to the client and write `prepared` status on the coordinator then we go and write commit marker on all participating data partitions.